google/cloud/examples/gcs2cbt.cc

```diff
@@ -0,0 +1,343 @@
+  }
+
+  void Push(T data) {
+    std::unique_lock<std::mutex> lk(mu_);
+    cv_.wait(lk, [this]() { return not Full(); });
+    buffer_[tail_] = std::move(data);
+    ++tail_;
+    if (tail_ >= buffer_.size()) {
+      tail_ = 0;
+    }
+    empty_ = false;
+    lk.unlock();
+    cv_.notify_all();
+  }
+
+ private:
+  bool Empty() { return head_ == tail_ and empty_; }
+  bool Full() { return head_ == tail_ and not empty_; }
+
+ private:
+  std::mutex mu_;
+  std::condition_variable cv_;
+  std::vector<T> buffer_;
+  std::size_t head_;
+  std::size_t tail_;
+  bool empty_;
+  bool is_shutdown_;
+};
+
+/**
+ * Breakdown a CSV file line into fields.
+ *
+ * TODO() - handle escape sequences with backslash.
+ *
+ * TODO() - handle quoted fields with embedded separators.
+ *
+ * TODO() - make the separator configurable.
+ */
+std::vector<std::string> ParseLine(long lineno, std::string const& line,
+                                   char separator);
+
+using CircularBuffer = GenericCircularBuffer<cbt::BulkMutation>;
+
+struct Options {
+  char separator;
+  std::vector<int> keys;
+  std::string keys_separator;
+
+  std::string ConsumeArg(int& argc, char* argv[], char const* arg_name) {
+    std::string const separator_option = "--separator=";
+    std::string const key_option = "--key=";
+    std::string const keys_separator_option = "--key-separator=";
+
+    std::string const usage = R""(
+[options] <project> <instance> <table> <family> <bucket> <object>
+The options are:
+    --help: produce this help.
+    --separator=c: use the 'c' character instead of comma (',') to separate the
+        values in the CSV file.
+    --key=N: use field number N as part of the row key. The fields are numbered
+        starting at one. They are concatenated in the order provided.
+    --key-separator=sep: use 'sep' to separate the fields when forming the row
+        key.
+    project: the Google Cloud Platform project id for your table.
+    instance: the Cloud Bigtable instance hosting your table.
+    table: the table where you want to upload the CSV file.
+    family: the column family where you want to upload the CSV file.
+    bucket: the name of the GCS bucket that contains the data.
+    object: the name of the GCS object that contains the data.
+)"";
+    while (argc >= 2) {
+      std::string argument(argv[1]);
+      std::copy(argv + 2, argv + argc, argv + 1);
+      argc--;
+      if (argument == "--help") {
+      } else if (0 == argument.find(separator_option)) {
+        separator = argument.substr(separator_option.size())[0];
+      } else if (0 == argument.find(key_option)) {
+        keys.push_back(std::stoi(argument.substr(key_option.size())) - 1);
+      } else if (0 == argument.find(keys_separator_option)) {
+        keys_separator = argument.substr(keys_separator_option.size());
+      } else {
+        return argument;
+      }
+    }
+    std::string cmd = argv[0];
+    auto last_slash = std::string(cmd).find_last_of('/');
```

Comment:
I assume this is to reduce something like `/path/to/executable` to `executable`?  If so, this won't achieve the desired result on Windows where the pathname component separator is `\`.  (Non-issue if this doesn't even work for Windows - I have no context on the Bigtable stuff)